lib/ssh/src/ssh.hrl

```diff
@@ -1307,5 +1307,75 @@ in the User's Guide chapter.
+
+-define(SSH_ERROR_REPORT(FMT, ARGS), logger:log(error, ?ERROR_REPORT_FUN(FMT), ARGS)).
+-define(SSH_WARNING_REPORT(FMT, ARGS), logger:log(warning, ?WARNING_REPORT_FUN(FMT), ARGS)).
+%% notice user with INFO_REPORT_FUN because that is how error_logger behaves
+-define(SSH_NOTICE_REPORT(FMT, ARGS), logger:log(notice, ?INFO_REPORT_FUN(FMT), ARGS)).
+%% help macros
+-define(ERROR_REPORT_FUN(FMT), ?REPORT_FUN(FMT, error_report, std_error, error_report)).
+-define(WARNING_REPORT_FUN(FMT), ?REPORT_FUN(FMT, warning_report, std_warning, warning_report)).
+-define(INFO_REPORT_FUN(FMT), ?REPORT_FUN(FMT, info_report, std_info, info_report)).
+%% macro below retains the original log event structure generated by error_logger
+-define(REPORT_FUN(FMT, TAG, TYPE, FUNC),
+        begin
+            fun(__Args0) ->
+                    __Args = [begin
+                                  case is_function(__Arg) of
+                                      true ->
+                                          __Arg();
+                                      _ ->
+                                          __Arg
+                                  end
+                              end || __Arg <- __Args0],
+                    __Report =
+                        #{label => {error_logger, FUNC},
+                          report =>
+                              %% FIXME should flatten be called here?
+                              lists:flatten(io_lib:format(FMT, __Args))},
+                    __Meta =
+                        #{error_logger =>
+                              #{tag => TAG, type => TYPE},
+                          report_cb => ?REPORT_TO_FORMAT_FUN},
+                    {__Report, __Meta}
+            end
+        end).
+%% -----------------------------------------------------------------
+%% Report formatting - i.e. Term => {Format,Args}
+%% This was earlier done in the event handler (error_logger_tty_h, etc)
+%% -----------------------------------------------------------------
+-define(REPORT_TO_FORMAT_FUN,
+        begin
+            fun(#{label:={error_logger,_},
+                  report:=___Report}) when is_map(___Report) ->
+                    %% logger:format_otp_report does maps:to_list, and for backwards
+                    %% compatibility reasons we don't want that.
+                    {"~tp\n",[___Report]};
+               (#{label:={error_logger,_},
+                  format:=___Format,
+                  args:=___Args}) ->
+                    %% This is not efficient, but needed for backwards compatibility
+                    %% in giving faulty arguments to the *_msg functions.
+                    try io_lib:scan_format(___Format,___Args) of
+                        _ -> {___Format,___Args}
+                    catch _:_ ->
+                            {"ERROR: ~tp - ~tp",[___Format,___Args]}
+                    end;
+               (___Term) ->
+                    logger:format_otp_report(___Term)
```

Comment:
`looger:format_otp_report/1` is an undocumented, internal API that belongs to another application. Should that really be used here?